android/app/src/main/kotlin/net/mullvad/mullvadvpn/viewmodel/SplashViewModel.kt

```diff
@@ -0,0 +1,110 @@
+package net.mullvad.mullvadvpn.viewmodel
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.selects.onTimeout
+import kotlinx.coroutines.selects.select
+import net.mullvad.mullvadvpn.lib.ipc.Event
+import net.mullvad.mullvadvpn.lib.ipc.MessageHandler
+import net.mullvad.mullvadvpn.lib.ipc.events
+import net.mullvad.mullvadvpn.model.AccountAndDevice
+import net.mullvad.mullvadvpn.model.AccountExpiry
+import net.mullvad.mullvadvpn.model.DeviceState
+import net.mullvad.mullvadvpn.repository.DeviceRepository
+import net.mullvad.mullvadvpn.repository.PrivacyDisclaimerRepository
+
+class SplashViewModel(
+    private val privacyDisclaimerRepository: PrivacyDisclaimerRepository,
+    private val deviceRepository: DeviceRepository,
+    private val messageHandler: MessageHandler,
+) : ViewModel() {
+
+    private val _uiSideEffect =
+        MutableSharedFlow<SplashUiSideEffect>(replay = 1, extraBufferCapacity = 1)
+    val uiSideEffect = _uiSideEffect.asSharedFlow()
+
+    fun start() {
+        viewModelScope.launch {
+            if (privacyDisclaimerRepository.hasAcceptedPrivacyDisclosure()) {
+                _uiSideEffect.emit(getStartDestination())
+            } else {
+                _uiSideEffect.emit(SplashUiSideEffect.NavigateToPrivacyDisclaimer)
+            }
+        }
+    }
+
+    private suspend fun getStartDestination(): SplashUiSideEffect {
+        val deviceState =
```

Comment:
Just a thought, but is it possible if we never get a device state that the user will be stuck in the splash screen forever? Is this a likely/possible scenario?

Reply:
Good question, that would basically mean we don't have the daemon right? Not sure but let's discuss offline how we should handle this. This remind me, we should also make sure we leave the app if daemon dies. Don't think we do it correctly right now.